### ISSUE TITLE
4.22 Build failed

### ISSUE THREAD

**Issue by saadat99** [npruehs/ue4-rts#103]
Hi, this is on the release branch, UE4.22
how bad these errors are?
```
Severity	Code	Description	Project	File	Line	Suppression State
Error (active)	E1850	cannot override 'final' function "AController::Possess" (declared at line 266 of "C:\Program Files\Epic Games\UE_4.22\Engine\Source\Runtime\Engine\Classes\GameFramework\Controller.h")	RTSPlugin	C:\Users\sam\Documents\Unreal Projects\RTSPlugin\Plugins\RealTimeStrategy\Source\RealTimeStrategy\Classes\RTSPlayerAIController.h	57	
Error	C3248	'AController::Possess': function declared as 'final' cannot be overridden by 'ARTSPlayerAIController::Possess'	RTSPlugin	C:\Users\sam\Documents\Unreal Projects\RTSPlugin\Plugins\RealTimeStrategy\Source\RealTimeStrategy\Classes\RTSPlayerAIController.h	57	
Error	C3248	'AController::Possess': function declared as 'final' cannot be overridden by 'ARTSPawnAIController::Possess'	RTSPlugin	C:\Users\sam\Documents\Unreal Projects\RTSPlugin\Plugins\RealTimeStrategy\Source\RealTimeStrategy\Classes\RTSPawnAIController.h	75	
Error	C3248	'AController::Possess': function declared as 'final' cannot be overridden by 'ARTSPawnAIController::Possess'	RTSPlugin	C:\Users\sam\Documents\Unreal Projects\RTSPlugin\Plugins\RealTimeStrategy\Source\RealTimeStrategy\Classes\RTSPawnAIController.h	75	
Error	C3248	'AController::Possess': function declared as 'final' cannot be overridden by 'ARTSPawnAIController::Possess'	RTSPlugin	C:\Users\sam\Documents\Unreal Projects\RTSPlugin\Plugins\RealTimeStrategy\Source\RealTimeStrategy\Classes\RTSPawnAIController.h	75	
Error	C3248	'AController::Possess': function declared as 'final' cannot be overridden by 'ARTSPlayerAIController::Possess'	RTSPlugin	C:\Users\sam\Documents\Unreal Projects\RTSPlugin\Plugins\RealTimeStrategy\Source\RealTimeStrategy\Classes\RTSPlayerAIController.h	57	
Error	MSB3075	The command ""C:\Program Files\Epic Games\UE_4.22\Engine\Build\BatchFiles\Build.bat" RTSPluginEditor Win64 Development -Project="C:\Users\sam\Documents\Unreal Projects\RTSPlugin\RTSPlugin.uproject" -WaitMutex -FromMsBuild" exited with code 5. Please verify that you have sufficient rights to run this command.	RTSPlugin	C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.MakeFile.Targets	44	

```
Full output
```1>------ Skipped Build: Project: UE4, Configuration: BuiltWithUnrealBuildTool Win32 ------
1>Project not selected to build for this solution configuration 
2>------ Build started: Project: RTSPlugin, Configuration: Development_Editor x64 ------
2>Parsing headers for RTSPluginEditor
2>  Running UnrealHeaderTool "C:\Users\sam\Documents\Unreal Projects\RTSPlugin\RTSPlugin.uproject" "C:\Users\sam\Documents\Unreal Projects\RTSPlugin\Intermediate\Build\Win64\RTSPluginEditor\Development\RTSPluginEditor.uhtmanifest" -LogCmds="loginit warning, logexit warning, logdatabase error" -Unattended -WarningsAsErrors -installed
2>Reflection code generated for RTSPluginEditor in 4.8603605 seconds
2>Using Visual Studio 2019 14.21.27702 toolchain (C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.21.27702) and Windows 10.0.17763.0 SDK (C:\Program Files (x86)\Windows Kits\10).
2>Building 9 actions with 8 processes...
2>  [1/9] Module.RealTimeStrategy.gen.3_of_4.cpp
2>C:\Users\sam\Documents\Unreal Projects\RTSPlugin\Plugins\RealTimeStrategy\Source\RealTimeStrategy/Classes/RTSPlayerAIController.h(57): error C3248: 'AController::Possess': function declared as 'final' cannot be overridden by 'ARTSPlayerAIController::Possess'
2>  C:\Program Files\Epic Games\UE_4.22\Engine\Source\Runtime\Engine\Classes\GameFramework/Controller.h(266): note: see declaration of 'AController::Possess'
2>  [2/9] Module.RealTimeStrategyTests.cpp
2>C:\Users\sam\Documents\Unreal Projects\RTSPlugin\Plugins\RealTimeStrategy\Source\RealTimeStrategy\Classes\RTSPawnAIController.h(75): error C3248: 'AController::Possess': function declared as 'final' cannot be overridden by 'ARTSPawnAIController::Possess'
2>  C:\Program Files\Epic Games\UE_4.22\Engine\Source\Runtime\Engine\Classes\GameFramework/Controller.h(266): note: see declaration of 'AController::Possess'
2>  [3/9] Module.RealTimeStrategy.gen.2_of_4.cpp
2>C:\Users\sam\Documents\Unreal Projects\RTSPlugin\Plugins\RealTimeStrategy\Source\RealTimeStrategy/Classes/RTSPawnAIController.h(75): error C3248: 'AController::Possess': function declared as 'final' cannot be overridden by 'ARTSPawnAIController::Possess'
2>  C:\Program Files\Epic Games\UE_4.22\Engine\Source\Runtime\Engine\Classes\GameFramework/Controller.h(266): note: see declaration of 'AController::Possess'
2>  [4/9] Module.RealTimeStrategy.cpp
2>C:\Users\sam\Documents\Unreal Projects\RTSPlugin\Plugins\RealTimeStrategy\Source\RealTimeStrategy\Classes\RTSPawnAIController.h(75): error C3248: 'AController::Possess': function declared as 'final' cannot be overridden by 'ARTSPawnAIController::Possess'
2>  C:\Program Files\Epic Games\UE_4.22\Engine\Source\Runtime\Engine\Classes\GameFramework/Controller.h(266): note: see declaration of 'AController::Possess'
2>C:\Users\sam\Documents\Unreal Projects\RTSPlugin\Plugins\RealTimeStrategy\Source\RealTimeStrategy\Classes\RTSPlayerAIController.h(57): error C3248: 'AController::Possess': function declared as 'final' cannot be overridden by 'ARTSPlayerAIController::Possess'
2>  C:\Program Files\Epic Games\UE_4.22\Engine\Source\Runtime\Engine\Classes\GameFramework/Controller.h(266): note: see declaration of 'AController::Possess'
2>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.MakeFile.Targets(44,5): error MSB3075: The command ""C:\Program Files\Epic Games\UE_4.22\Engine\Build\BatchFiles\Build.bat" RTSPluginEditor Win64 Development -Project="C:\Users\sam\Documents\Unreal Projects\RTSPlugin\RTSPlugin.uproject" -WaitMutex -FromMsBuild" exited with code 5. Please verify that you have sufficient rights to run this command.
2>Done building project "RTSPlugin.vcxproj" -- FAILED.
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 1 skipped ==========
```

**Comment by salvationws** [comment]
it will be good to update this project to 4.22 :)

**Comment by npruehs** [comment]
Hey everyone,

I haven't installed 4.22 on my machine (yet), but you should be fine with overriding OnPossessed instead :)

**Comment by grassandberries** [comment]
I'm having the same issue. How do you override OnPossess???

**Comment by ash501** [comment]
Hello,
I just compiled the plugin and a test project in 4.22.3

To override OnPossess... check RTSCharacterAIController.h/.cpp and RTSPlayerAIController.h/cpp in both files you'll find the function causing the error (ARTSPlayerAIController::Possess). Rename this function to  ARTSCharacterAIController::OnPossess (and ARTSPlayerAIController::OnPossess in the other cpp). In the header files (.h) rename  virtual void Possess(APawn* InPawn) override; to  virtual void OnPossess(APawn* InPawn) override;

Now, the plugin should compile. I'll run some functional tests later.
@npruehs If you want to have the modified files, let me know.

So long and God bless you,

**Comment by saadat99** [comment]
> Hello,
> I just compiled the plugin and a test project in 4.22.3
> 
> To override OnPossess... check RTSCharacterAIController.h/.cpp and RTSPlayerAIController.h/cpp in both files you'll find the function causing the error (ARTSPlayerAIController::Possess). Rename this function to ARTSCharacterAIController::OnPossess (and ARTSPlayerAIController::OnPossess in the other cpp). In the header files (.h) rename virtual void Possess(APawn* InPawn) override; to virtual void OnPossess(APawn* InPawn) override;
> 
> Now, the plugin should compile. I'll run some functional tests later.
> @npruehs If you want to have the modified files, let me know.
> 
> So long and God bless you,

**Comment by ash501** [comment]
It crashes here with a Null-Pointer
AttackComponent = InPawn->FindComponentByClass<URTSAttackComponent>();

Looking into it, but haven't found anything yet. It crashes basically on any object inherit from this class.

**Comment by saadat99** [comment]
Any success on this guys? I've been waiting so long to make this work

**Comment by grassandberries** [comment]
Does anyone have this project compiled successfully for UE4 (4.22) build?

**Comment by npruehs** [comment]
Hey @saadat99 @grassandberries the steps outlined above should fix the issue, as mentioned before. I haven't got UE 4.22 on my machine yet (our project is picking up new engine versions slowly to make sure nothing breaks), but I'll update this thread as soon as I was able to push the fixes.

**Comment by npruehs** [comment]
Fixed in commit 6dfa12de6ee60f50b97741105012e8b7b83c3762.

**Comment by ash501** [comment]
Thank you @npruehs.
I confirm that it works. My last issue from yesterday was that I missed the configuration. All seems to be fine now.
